lib/connector/OAuthConnector.js

```diff
@@ -182,52 +227,90 @@ class OAuthConnector {
+     * For the vendor's system, if the currently stored access token is expired or nearing expiry, and a refresh token is available, a new access
      * token is obtained, stored for future use, and returned. If a current access token cannot be returned, an exception is thrown.
      * @param {FusebitContext} fusebitContext The Fusebit context of the request
      * @param {*} userContext The vendor user context
+     * @param {string} foreignVendorId If specified, gets a valid access token for the OAuth connector identified by the
+     * foreignVendorId entry in the userContext.foreignOAuthIdentities rather than a user of this connector.
      */
-    async ensureAccessToken(fusebitContext, userContext) {
-        if (userContext.vendorToken.access_token && userContext.vendorToken.expires_at > Date.now() + 30000) {
-            Sdk.debug('RETURNING CURRENT ACCESS TOKEN FOR USER', userContext.vendorUserId);
-            return userContext.vendorToken;
-        }
-        if (userContext.vendorToken.refresh_token) {
-            Sdk.debug('REFRESHING ACCESS TOKEN FOR USER', userContext.vendorUserId);
-            userContext.status = 'refreshing';
-            try {
-                await this.saveUser(fusebitContext, userContext);
-                userContext.vendorToken = await this.refreshAccessToken(fusebitContext, userContext.vendorToken);
-                if (!isNaN(userContext.vendorToken.expires_in)) {
-                    userContext.vendorToken.expires_at = Date.now() + +userContext.vendorToken.expires_in * 1000;
-                }
-                userContext.vendorUserProfile = await this.getUserProfile(userContext.vendorToken);
-                userContext.status = 'authenticated';
-                await this.saveUser(fusebitContext, userContext);
+    async ensureAccessToken(fusebitContext, userContext, foreignVendorId) {
+        if (foreignVendorId) {
+            // Get access token from foreign OAuth connector specified in userContext.foreignOAuthIdentities
+            const oauthIdentity = (userContext.foreignOAuthIdentities || {})[foreignVendorId];
+            Sdk.debug('OBTAINING ACCESS TOKEN FOR FOREIGN USER', foreignVendorId, oauthIdentity);
+            if (oauthIdentity) {
+                const response = await Superagent.get(
+                    `${oauthIdentity.connectorBaseUrl}/user/${encodeURIComponent(oauthIdentity.userId)}/token`
+                ).set('Authorization', `Bearer ${fusebitContext.fusebit.functionAccessToken}`);
+                return response.body;
+            } else {
+                throw new Error(
+                    `The user ${userContext.vendorUserId} is not associated with an identity in the ${foreignVendorId} OAuth connector.`
+                );
+            }
+        } else {
+            // Get access token for "this" OAuth connector
+            if (userContext.vendorToken.access_token && userContext.vendorToken.expires_at > Date.now() + 30000) {
```

Comment:
Magic constant.

Reply:
Done. LIfted to public class field so that instances/derived classes can change as needed.